### PR TITLE
[Ability] (Partially) Implement synchronize ability with old psycho shift interaction

### LINKED ISSUE #2694: [Feature] Implement Synchronize
**Is your feature request related to a problem? Please describe.**
Synchronize is cursed, every time someone wants to implement it, they either disappear or they see #872 and they desist, thinking it's almost done... but it has no progress for a month, the game has changed a lot so I doubt this is going to be merged

**Describe the Feature**
Someone please take over and implement Synchronize to the game

**Additional context**
From Bulbapedia: 
When a Pokémon with Synchronize is burned, paralyzed, or poisoned by another Pokémon, that Pokémon will be inflicted with the same status condition. Synchronize will not inflict confusion, sleep, or freeze.

Even if Synchronize would have no effect due to the other Pokémon already having a status condition or the other Pokémon's type, Ability, or team condition (e.g. Safeguard), it will still activate (revealing it as the Pokémon's Ability), but have no effect. Synchronize can affect a Pokémon behind a substitute. Synchronize can be activated by other Pokémon's Abilities, such as Static, but cannot be activated by the Toxic Orb or Flame Orb.

If a Pokémon with Synchronize has a Berry that cures the status condition it was just inflicted with, Synchronize will activate before the Berry is eaten.

If Psycho Shift is used by a burned, paralyzed, poisoned, or badly poisoned Pokémon targeting a Pokémon with Synchronize, Synchronize now activates before Psycho Shift cures the user of its status condition. Therefore, Synchronize will fail to inflict the status condition anew in such a case.

If multiple Pokémon receive a status from the same move, Synchronize now gets an opportunity to activate for any Pokémon that has it, regardless of what order they received the status in. Synchronize can now cause another Pokémon's Synchronize to activate, but if this happens, the original Pokémon therefore already has a status condition, and it will merely cause an ability message to appear with no other effect.

Toxic Spikes do not activate this Ability unless a Pokémon on the field directly causes a Pokémon with Synchronize to switch in, such as by using the moves Roar or Whirlwind.

There are some pokemon that has Synchronize as their only ability, like Mew, and Mew... and... Mew, poor Mew, please help Mew, thanks

### PR DESCRIPTION
**Pull request by yilinggk**
## What are the changes?
Adds (partial) implementation for Synchronize ability as per https://github.com/pagefaultgames/pokerogue/issues/2694. Marking the ability as Partial since this implementation interacts with Psycho Shift the way it was intended before Gen V. Ie:
1. Pyscho shift clears user of status and then transfers to opponent
2. Opponent has synchronize and send the status back to user

There were some race issues with promises that caused buggy behavior with this interaction if we want to implement it the way it was intended post Gen V. That will need more time and code changes to debug, thus doing this for now.

Closes #2694 

## Why am I doing these changes?
More details here: https://github.com/pagefaultgames/pokerogue/issues/2694

## What did change?

- Added new postSetStatusAblAttr ability hook that calls to synchronize status with opposing Pokemon
- Updated Psycho shift logic to clear user status first and then set status on opponent. This shouldn't affect any scenarios other than synchronize.

### Screenshots/Videos
Synchronize working:
syncInteract

Psycho shift interaction (old pre gen V):
psychoShiftInteract

## How to test the changes?
Added new UTs for the ability and tested manually with overrides on localhost. See videos above. Please set variables in overrides.ts

## Checklist
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

**Comment by flx-sta** [comment]
@yilinggk please make sure to link the issues it would close too via ["magic" keywords](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests#linking-a-pull-request-to-an-issue)

**Comment by yilinggk** [comment]
> @yilinggk please make sure to link the issues it would close too via ["magic" keywords](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests#linking-a-pull-request-to-an-issue)

done, thanks!

**Comment by yilinggk** [comment]
> #872 and #2694 may be relevant

This change builds off of #872 since that PR has been abandoned since March as mentioned in #2694.

File: src/field/pokemon.ts [hunk 1]
```diff
@@ -23,7 +23,7 @@ import { BattlerTag, BattlerTagLapseType, EncoreTag, GroundedTag, HelpingHandTag
...
+import { Ability, AbAttr, BattleStatMultiplierAbAttr, BlockCritAbAttr, BonusCritAbAttr, BypassBurnDamageReductionAbAttr, FieldPriorityMoveImmunityAbAttr, IgnoreOpponentStatChangesAbAttr, MoveImmunityAbAttr, MoveTypeChangeAttr, PreApplyBattlerTagAbAttr, PreDefendFullHpEndureAbAttr, ReceivedMoveDamageMultiplierAbAttr, ReduceStatusEffectDurationAbAttr, StabBoostAbAttr, StatusEffectImmunityAbAttr, TypeImmunityAbAttr, VariableMovePowerAbAttr, WeightMultiplierAbAttr, allAbilities, applyAbAttrs, applyBattleStatMultiplierAbAttrs, applyPreApplyBattlerTagAbAttrs, applyPreAttackAbAttrs, applyPreDefendAbAttrs, applyPreSetStatusAbAttrs, UnsuppressableAbilityAbAttr, SuppressFieldAbilitiesAbAttr, NoFusionAbilityAbAttr, MultCritAbAttr, IgnoreTypeImmunityAbAttr, DamageBoostAbAttr, IgnoreTypeStatusEffectImmunityAbAttr, ConditionalCritAbAttr, applyFieldBattleStatMultiplierAbAttrs, FieldMultiplyBattleStatAbAttr, AllyMoveCategoryPowerBoostAbAttr, FieldMoveTypePowerBoostAbAttr, AddSecondStrikeAbAttr, UserFieldMoveTypePowerBoostAbAttr, PostSetStatusAbAttr, applyPostSetStatusAbAttrs } from "../data/ability";
```
Comment (flx-sta) [inline src/field/pokemon.ts:26]: Little hard to see here what the change is.

File: src/test/abilities/synchronize.test.ts
Comment (flx-sta) [inline src/test/abilities/synchronize.test.ts:1]: do these tests address held berry RNG and ability RNG that might block the sync?

**Comment by DayKev** [comment]
Are you still around to update this PR?

**Comment by DayKev** [comment]
Oops, forgot to convert to draft earlier.

**Comment by yilinggk** [comment]
> Are you still around to update this PR?

@DayKev I can make updates if necessary! What does this need before we can get it checked in?

File: src/field/pokemon.ts (hunk 1 above)
Comment (yilinggk) [reply src/field/pokemon.ts]: this is just adding a few new imports. Not much I can do about the formatting if we want to keep it consistent with the rest of the code

**Comment by DayKev** [comment]
At the moment we should probably double check that Psycho Shift/Synchronize are still required to work based on old behavior.

**Comment by DayKev** [comment]
Actually the `simulate` property also needs to be added (for AI use).

**Review by flx-sta — DISMISSED** [review]

File: src/data/ability.ts [hunk 2]
```diff
@@ -1785,6 +1785,61 @@ export class PostDefendStealHeldItemAbAttr extends PostDefendAbAttr {
...
+   * If the `StatusEffect` that was set is Burn, Paralysis, Poison, or Toxic, and the status
+   * was set by a source Pokemon, set the source Pokemon's status to the same `StatusEffect`.
+   * @param pokemon {@linkcode Pokemon} that status condition was set on.
+   * @param sourcePokemon {@linkcode Pokemon} that that set the status condition. Is null if status was not set by a Pokemon.
+   * @param passive Whether this ability is a passive.
+   * @param effect {@linkcode StatusEffect} that was set.
+   * @param args Set of unique arguments needed by this attribute.
+   * @returns `true` if application of the ability succeeds.
+   */
+  override applyPostSetStatus(pokemon: Pokemon, sourcePokemon: Pokemon | null = null, passive: boolean, effect: StatusEffect, simulated:boolean, args: any[]): boolean {
```
Comment (innerthunder) [inline src/data/ability.ts:1823]:
```suggestion
  override applyPostSetStatus(pokemon: Pokemon, sourcePokemon: Pokemon | null = null, passive: boolean, effect: StatusEffect, simulated: boolean, args: any[]): boolean {
```

File: src/data/ability.ts [hunk 3]
```diff
@@ -4664,6 +4719,10 @@ export function applyStatMultiplierAbAttrs(attrType: Constructor<StatMultiplierA
   pokemon: Pokemon, stat: BattleStat, statValue: Utils.NumberHolder, simulated: boolean = false, ...args: any[]): Promise<void> {
   return applyAbAttrsInternal<StatMultiplierAbAttr>(attrType, pokemon, (attr, passive) => attr.applyStatStage(pokemon, passive, simulated, stat, statValue, args), args);
 }
+export function applyPostSetStatusAbAttrs(attrType: Constructor<PostSetStatusAbAttr>,
+  pokemon: Pokemon, effect: StatusEffect, sourcePokemon?: Pokemon | null, simulated: boolean = false, ...args: any[]): Promise<void> {
+  return applyAbAttrsInternal<PostSetStatusAbAttr>(attrType, pokemon, (attr, passive) => attr.applyPostSetStatus(pokemon, sourcePokemon, passive, effect, simulated, args), args);
```
Comment (innerthunder) [inline src/data/ability.ts:4724]:
```suggestion
  return applyAbAttrsInternal<PostSetStatusAbAttr>(attrType, pokemon, (attr, passive) => attr.applyPostSetStatus(pokemon, sourcePokemon, passive, effect, simulated, args), args, false, simulated);
```
The ability popup might show at an incorrect time without this

**Review by innerthunder — DISMISSED** [review]
Slight issue with how simulated calls are handled

**Review by DayKev — APPROVED** [review]
🤔

**Review by innerthunder — APPROVED** [review]

**Review by flx-sta — APPROVED** [review]

**Pull request merged**